### PR TITLE
refactor: Consolidate chart logic

### PR DESCRIPTION
**Pull request by bprusinowski**
This PR consolidates common chart-state-computing logic by extracting some chart logic outside of individual chart states:
- computation of data needed by charts (`chartData`, `scalesData`, `segmentData`, `allData`). I introduced this distinction to make all charts "ready for interactive filters" (e.g. maps do not have them yet) and to have one place to "fix" things, instead of needing to go to every chart,
- sorting the data. Right now every chart type defines its own sorting / missing values filtering logic and this information is used when computing the data upfront, before it's passed to the chart.

It also fixes some minor things:
- column exit animations,
- inner band scale when toggling interactive legend filters,
- tests with dependencies that import `rdf-js` modules (_SyntaxError: Cannot use import statement outside a module_),
- scales now adapt to color legend filter when using time slider,
- it's now not possible to remove all color legend filters to prevent showing no chart.

### Reference: Interactive filters
- Time range
- Time slider
- Color legend

### New chart data structure
- `chartData`: data with all interactive filters applied, used to draw the shapes,
- `scalesData`: data with color legend and time range filters applied, used to compute the scales. As time slider filter is not applied to it, it holds the information about all potential scale values and makes it possible to fix the scales when animating the charts,
- `segmentData`: data with time range filter applied. Needed to hold the information about all color values that can be used in a chart, to save unchecked color legend items from being removed,
- `allData`: full dataset, needed to show the timeline when using time slider. We can't use `scalesData` here, due to the fact the the `useChartData` hook gets re-rendered and prevents the timeline from working it such case.

### Next steps
- Consolidate segments computation logic.

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -1,28 +1,42 @@
 {
   "extends": "next",
-  "plugins": ["unused-imports", "import", "visualize-admin"],
+  "plugins": ["import", "visualize-admin"],
```
Comment (ptbrowne) [inline .eslintrc.json:3]: Just for my understanding: the auto fix still removes unused imports ?

File: app/domain/data.ts [hunk 2]
```diff
@@ -228,31 +228,31 @@ export const isGeoDimension = (
...
 export const isOrdinalMeasure = (
-  dimension?: DimensionMetadataFragment
+  dimension?: { __typename?: string } | null
 ): dimension is OrdinalMeasure => {
   return dimension?.__typename === "OrdinalMeasure";
 };
 
 export const isTemporalDimension = (
-  dimension?: DimensionMetadataFragment | null
+  dimension?: { __typename?: string } | null
```
Comment (ptbrowne) [inline app/domain/data.ts:255]: Why this change ? Isn't it a bit less semantic ?

File: .eslintrc.json (hunk 1 above)
Comment (bprusinowski) [reply .eslintrc.json:3]: yes :)

File: app/charts/shared/chart-state.ts [hunk 3]
```diff
@@ -1,8 +1,23 @@
...
+export type ChartStateMetadata = {
+  sortData: (data: Observation[]) => Observation[];
+};
+
+export const getChartStateMetadata = (
+  chartConfig: ChartConfig
+): ChartStateMetadata | undefined => {
+  switch (chartConfig.chartType) {
+    case "area":
+      return getAreasStateMetadata(chartConfig);
```
Comment (ptbrowne) [inline app/charts/shared/chart-state.ts]: By having a switch here, we might need to load all the getMetadata functions for all chart types (when we only need the getMetadata for 1 chart type, especially when viewing). I wonder if it would make sense to have a way so that the logic can be imported async.

File: app/charts/chart-loading-wrapper.tsx [hunk 4]
```diff
@@ -86,12 +100,12 @@ export const ChartLoadingWrapper = <
...
         <A11yTable
           title={title}
-          observations={observations.data}
+          observations={observations}
           dimensions={dimensions}
           measures={measures}
         />
         {React.createElement(Component, {
-          data: observations.data,
+          data: observations,
```
Comment (ptbrowne) [inline app/charts/chart-loading-wrapper.tsx]: If I am not mistaken, the Component here is loaded async; we could "attach" the getMetadata function a static method to it, so that it's loaded async.

File: app/charts/map/map-state.tsx [hunk 5]
```diff
@@ -578,6 +581,14 @@ const useMapState = (
   };
 };
 
+export const getMapStateMetadata = (): ChartStateMetadata => {
```
Comment (ptbrowne) [inline app/charts/map/map-state.tsx:593]: I think we could have the same method name (getMapStateMetadata -> getMetadata) across all the chart files so that we can load the module async and just call getMetadata on it. This  way, it is even clearer that a chart file needs to implement the getMetadata method.

File: app/charts/chart-loading-wrapper.tsx [hunk 6]
```diff
@@ -87,8 +87,17 @@ export const ChartLoadingWrapper = <
...
-  if (metadataData?.dataCubeByIri && dimensions && measures && data) {
-    const { title } = metadataData.dataCubeByIri;
+  // const { chartData, scalesData } = useChartData(data ?? [], {
+  //   interactiveFiltersConfig: chartConfig.interactiveFiltersConfig,
+  //   animationField: isAnimationInConfig(chartConfig)
+  //     ? chartConfig.fields.animation
+  //     : undefined,
+  //   getXDate: chartStateMetadata?.getXDate,
+  //   getSegment: chartStateMetadata?.getSegment,
+  // });
```
Comment (ptbrowne) [inline app/charts/chart-loading-wrapper.tsx]: Is this to remove ?

File: app/charts/shared/chart-state.ts [hunk 7]
```diff
@@ -53,13 +55,13 @@ export const getChartStateMetadata = (
...
-      return getLinesStateMetadata(chartConfig);
+      return getLinesStateMetadata(chartConfig, observations, dimensions);
     case "map":
       return getMapStateMetadata();
     case "pie":
-      return getPieStateMetadata();
+      return getPieStateMetadata(chartConfig, observations, dimensions);
     case "scatterplot":
-      return getScatterplotStateMetadata();
+      return getScatterplotStateMetadata(chartConfig, observations, dimensions);
```
Comment (ptbrowne) [inline app/charts/shared/chart-state.ts:69]: So I guess that if we have the same getMetadata method everywhere, we can remove the switch.

File: app/charts/shared/ChartProps.tsx [hunk 8]
```diff
@@ -3,7 +3,8 @@ import { Observation } from "@/domain/data";
 import { DimensionMetadataFragment } from "@/graphql/query-hooks";
 
 export type BaseChartProps = {
-  data: Observation[];
+  chartData: Observation[];
+  scalesData: Observation[];
   dimensions: DimensionMetadataFragment[];
   measures: DimensionMetadataFragment[];
 };
```
Comment (ptbrowne) [inline app/charts/shared/ChartProps.tsx:16]: Here we could have a little `/** comment */` with what you've written in PR description, so that when we are not sure what is scalesData, we can hover over it and see it :)

File: app/charts/column/columns-stacked.tsx [hunk 9]
```diff
@@ -18,40 +13,32 @@ export const ColumnsStacked = () => {
...
+          width: bandwidth,
+          height: yScale(segment[0]) - yScale(segment[1]),
+          color,
+        };
+      });
     });
   }, [bandwidth, colors, getX, series, xScale, yScale]);
 
   React.useEffect(() => {
+    console.log("rendering columns", renderData);
```
Comment (ptbrowne) [inline app/charts/column/columns-stacked.tsx:37]: 👁️

File: app/charts/column/columns-stacked.tsx [hunk 10]
```diff
@@ -18,40 +13,32 @@ export const ColumnsStacked = () => {
...
+
+        return {
+          key: `${key}-${x}`,
+          x: xScale(x) as number,
+          y: yScale(segment[1]),
+          width: bandwidth,
+          height: yScale(segment[0]) - yScale(segment[1]),
+          color,
+        };
+      });
```
Comment (ptbrowne) [inline app/charts/column/columns-stacked.tsx:32]: 👍

File: app/jest.config.js [hunk 11]
```diff
@@ -12,6 +12,10 @@ const config = {
     "^@/(.*)$": "<rootDir>/$1",
     "\\.(css)$": "<rootDir>/test/style-mock.js",
   },
+  transform: {
+    "node_modules/@rdf*": "ts-jest",
+  },
+  transformIgnorePatterns: ["node_modules/!@rdf*"],
```
Comment (ptbrowne) [inline app/jest.config.js:18]: 👏

**Review by ptbrowne — APPROVED** [review]
👏  🚀 

So nice to have found all this different "data" :) I have some comments here and there and I think the ability to import async is important, but overall very very nice 💯
